services/shhext/chat/protocol.go

```diff
@@ -0,0 +1,124 @@
+package chat
+
+import (
+	"crypto/ecdsa"
+	"errors"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/golang/protobuf/proto"
+)
+
+type ProtocolService struct {
+	log        log.Logger
+	encryption *EncryptionService
+	Enabled    bool
+}
+
+// NewProtocolService creates a new ProtocolService instance
+func NewProtocolService(encryption *EncryptionService) *ProtocolService {
+	return &ProtocolService{
+		log:        log.New("package", "status-go/services/sshext.chat"),
+		encryption: encryption,
+	}
+}
+
+func (p *ProtocolService) addBundleAndMarshal(myIdentityKey *ecdsa.PrivateKey, msg *ProtocolMessage) ([]byte, error) {
+	// Get a bundle
+	bundle, err := p.encryption.CreateBundle(myIdentityKey)
+	if err != nil {
+		p.log.Error("encryption-service", "error creating bundle", err)
+		return nil, err
+	}
+
+	msg.Bundle = bundle
+
+	// marshal for sending to wire
+	marshaledMessage, err := proto.Marshal(msg)
+	if err != nil {
+		p.log.Error("encryption-service", "error marshaling message", err)
+		return nil, err
+	}
+
+	return marshaledMessage, nil
+}
+
+// BuildPublicMessage marshals a public chat message given the user identity private key and a payload
+func (p *ProtocolService) BuildPublicMessage(myIdentityKey *ecdsa.PrivateKey, payload []byte) ([]byte, error) {
+	// Build message not encrypted
+	protocolMessage := &ProtocolMessage{
+		PublicMessage: payload,
+	}
+
+	return p.addBundleAndMarshal(myIdentityKey, protocolMessage)
+}
+
+// BuildDirectMessage marshals a 1:1 chat message given the user identity private key, the recipient's public key, and a payload
+func (p *ProtocolService) BuildDirectMessage(myIdentityKey *ecdsa.PrivateKey, theirPublicKeys []*ecdsa.PublicKey, payload []byte) (*map[*ecdsa.PublicKey][]byte, error) {
```

Comment:
Maps in Go are passed via a pointer already. `map[*ecdsa.PublicKey][]byte` should be fine.